### PR TITLE
adding mysql-to-kv templates

### PR DESCRIPTION
**Pull request by Sagikeren88**

### CODE REVIEW

File: mysql-to-kv/function.yaml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      value: rfamro
+    - type: value
+      name: SQL_DB_NAME
+      value: Rfam
+  runtime: "python:3.6"
+  minReplicas: 1
+  maxReplicas: 1
+  build:
+    commands:
+      - 'pip install requests pymysql pandas'
```
Comment (pavius) [inline mysql-to-kv/function.yaml]: No need for quotes

File: mysql-to-kv/function.yaml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+# limitations under the License.
+
+
+apiVersion: "nuclio.io/v1beta1"
+kind: "Function"
+spec:
+  description: "A simple v3io-webapi usage for PutItem using MySQL public DB"
+  handler: "main:handler"
+  env:
+    - type: value
```
Comment (pavius) [inline mysql-to-kv/function.yaml]: 1. No need for `type`
2. These values need to be passed as variables. See the values file in https://github.com/nuclio/nuclio-templates/tree/master/string-manipulator

File: mysql-to-kv/function.yaml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      name: SQL_DB_NAME
+      value: Rfam
+  runtime: "python:3.6"
+  minReplicas: 1
+  maxReplicas: 1
+  build:
+    commands:
+      - 'pip install requests pymysql pandas'
+    baseImage: "python:3.6-jessie"
```
Comment (pavius) [inline mysql-to-kv/function.yaml]: I would opt for `python:3.6` - it's used more often and also distributed to dark sites

File: mysql-to-kv/ingest_from_mysql.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+def _putitem(url, body, user, password):
+    """
+    PutItem request
+    """
+    headers = {'Content-Type': 'application/json', 'X-v3io-function': 'PutItem'}
+    res = requests.post(url=url, data=json.dumps(body), headers=headers, auth=HTTPBasicAuth(user, password))
+    return res
+
+
+def handler(context, event):
```
Comment (pavius) [inline mysql-to-kv/ingest_from_mysql.py]: 1. handle should be first function (convention)
2. you're creating a connection per invocation, use init_context (should be second function after handler - convention):

```
def init_context(context):
	
	# connect to database
	connection = ...

	# set connection as member of context
	setattr(context, 'connection', connection)
```

Then in your handler, simply use it:
```
context.connection.read_sql_query(...)
```

File: mysql-to-kv/ingest_from_mysql.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        port=int(port),
+        user=user,
+        passwd=password,
+        db=database,
+        charset='utf8mb4')
+    df = pd.read_sql_query('select rfam_acc,rfam_id,auto_wiki,description,author,seed_source FROM family LIMIT 10', conn)
+    for index, row in df.iterrows():
+        body = _pack_putitem_body(index, row)
+        try:
+            res = _putitem(f'http://v3io-webapi:8081/{container}/{table}/', body, igz_user, igz_password)
```
Comment (pavius) [inline mysql-to-kv/ingest_from_mysql.py]: The webapi hostname/port should also be a variable

File: mysql-to-kv/ingest_from_mysql.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        }
+    return body
+
+
+def _putitem(url, body, user, password):
+    """
+    PutItem request
+    """
+    headers = {'Content-Type': 'application/json', 'X-v3io-function': 'PutItem'}
+    res = requests.post(url=url, data=json.dumps(body), headers=headers, auth=HTTPBasicAuth(user, password))
```
Comment (pavius) [inline mysql-to-kv/ingest_from_mysql.py]: Why not just `return requests.post(...)`?

File: mysql-to-kv/ingest_from_mysql.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+
+def _pack_putitem_body(key, row):
+    """
+    PutItem request body formater
+    """
+    dict_item = dict()
+    for k, v in row.iteritems():
+        dict_item[k] = {'N' if isinstance(v, int) else 'S': '{}'.format(v)}
+    body = {
```
Comment (pavius) [inline mysql-to-kv/ingest_from_mysql.py]: do we need the body variable? 
```
return { 
    ...
}
```

File: mysql-to-kv/ingest_from_mysql.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    res = requests.post(url=url, data=json.dumps(body), headers=headers, auth=HTTPBasicAuth(user, password))
+    return res
+
+
+def handler(context, event):
+    #################
+    # IGZ variables #
+    #################
+    # table
+    table = os.getenv('TABLE', 'table')
```
Comment (pavius) [inline mysql-to-kv/ingest_from_mysql.py]: Once you move to variables, don't use defaults

File: mysql-to-kv/ingest_from_mysql.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        passwd=password,
+        db=database,
+        charset='utf8mb4')
+    df = pd.read_sql_query('select rfam_acc,rfam_id,auto_wiki,description,author,seed_source FROM family LIMIT 10', conn)
+    for index, row in df.iterrows():
+        body = _pack_putitem_body(index, row)
+        try:
+            res = _putitem(f'http://v3io-webapi:8081/{container}/{table}/', body, igz_user, igz_password)
+
+        except Exception:
```
Comment (pavius) [inline mysql-to-kv/ingest_from_mysql.py]: I wouldn't catch this exception. You're actually providing less information about what happened (the original exception will contain the root cause, and will result in returning 500 as well)

File: mysql-to-kv/ingest_from_mysql.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+            context.logger.warn("Failed to send request")
+            return context.Response(body='An issue was occured',
+                                headers={},
+                                content_type='text/plain',
+                                status_code=500)
+
+    return context.Response(body='PutItem completed successfully',
```
Comment (pavius) [inline mysql-to-kv/ingest_from_mysql.py]: No need to return text with 200 (machines don't read text). In fact no need to return anything, Nuclio will handle this as 204 where applicable IIRC

**Comment by Sagikeren88** [comment]
@pavius  - Thanks so much for the informative CR!
fixed the code per your comments + changed the following - 
* added a value YAML
* added init_context logic
* Used v3io-frames instead of webapi

**Pull request merged**